src/codecs/ico/decoder.rs

```diff
@@ -31,7 +31,9 @@ enum DecoderError {
     /// The enclosed PNG is not in RGBA, which is invalid: https://blogs.msdn.microsoft.com/oldnewthing/20101022-00/?p=12473/.
     PngNotRgba,
 
-    /// The optional mask row, containing 1 bit per pixel, padded to 4 bytes, was too short for this image.
+    /// This could be replaced with some error about an invalid image_length in the header
+    #[allow(dead_code)]
+    #[deprecated = "Because there is no encoding for the length of the mask this error corresponded to a different case than it claimed"]
     BmpIcoMaskTooShortForImage,
```

Comment:
This is a private enum so you can just remove unused variants. That said, should't it still check the length in case there is _some_ mask to ensure it's not only a partial one? Oh how I wish the test suite was larger..

Reply:
Oh yeah of course. I also see what you mean about the error. Do we want to attempt to load the ICO any way if the `image_length` is not one of the two expected values? If not I can make the check more strict again and remove the test case I added or keep it as an example of a malformed icon. Either way some fix is still needed for the decoder to avoid a panic.

This does raise a question: What is this library's attitude towards decoding strictness? I haven't looked at what this would mean for other codecs, but it would be useful for my application to be able to load the original logitech icon despite the malformed header. There are also some other `DecoderError` variants that are validation beyond what is necessary to load the image data. For example, some other icons I've come across use a technically invalid non-RGBA8 PNG color type, but are otherwise valid.

Reply:
As you can imagine, the mix of base crates for individual formats gives a range of approaches to the problem and there is no single consistent solution. I'll just explain how I would approach this. As a general guideline, I'd resist the temptation to guess. For technically unsupported color types there can hardly be any different semantics if it was allowed hence there isn't a lot of harm in supporting them.

However for partial masks this seems less clear. For too large values it seems okay again, it's also more compatible with previous releases which is an upside given the small test coverage. Since you found relatively consistent behaviour for empty masks, I'd definitely include that. Definitely keep the extra test here! GIMP, Chrome, Firefox, Sublime is also enough to say that this is more or less consistent where supported. Windows appears like a bug to me since it is apparently interpreting _some other_ data as a mask.. That's very fishy and uncontrollable.

Reply:
Ok, I've pushed another commit that I think should match the original strictness but report an error instead of panic on the invalid case in question.

How about making another constructor for `IcoDecoder` that has a bool strictness parameter? The default behavior of the decoder would be as it is now, but a user could construct an `IcoDecoder` directly to decode icons such as the logitech one? If that would be acceptable I could open another pull request.

Reply:
Unless you encounter an image in the while I wouldn't worry about the case too much. Plus I'd discuss it as part of another PR. The three basic options are a separate constructor or a bool field or a bool parameter.